File: tests/unit/src/com/android/settings/biometrics2/ui/model/EnrollmentRequestTest.java
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.settings.biometrics2.ui.model;

import static com.android.settings.biometrics.BiometricEnrollActivity.EXTRA_SKIP_INTRO;
import static com.android.settings.biometrics2.ui.model.EnrollmentRequest.EXTRA_SKIP_FIND_SENSOR;

import static com.google.android.setupcompat.util.WizardManagerHelper.EXTRA_IS_DEFERRED_SETUP;
import static com.google.android.setupcompat.util.WizardManagerHelper.EXTRA_IS_PORTAL_SETUP;
import static com.google.android.setupcompat.util.WizardManagerHelper.EXTRA_IS_SETUP_FLOW;
import static com.google.android.setupcompat.util.WizardManagerHelper.EXTRA_IS_SUW_SUGGESTED_ACTION_FLOW;
import static com.google.common.truth.Truth.assertThat;

import android.annotation.NonNull;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class EnrollmentRequestTest {

    @NonNull
    private final Context mContext = ApplicationProvider.getApplicationContext();

    @Test
    public void testIsSuw() {
        // Default false
        assertThat(new EnrollmentRequest(new Intent(), mContext, true).isSuw()).isFalse();
        assertThat(new EnrollmentRequest(new Intent(), mContext, false).isSuw()).isFalse();

        final Intent trueIntent = new Intent();
        trueIntent.putExtra(EXTRA_IS_SETUP_FLOW, true);
        assertThat(new EnrollmentRequest(trueIntent, mContext, true).isSuw()).isTrue();
        assertThat(new EnrollmentRequest(trueIntent, mContext, false).isSuw()).isFalse();

        final Intent falseIntent = new Intent();
        trueIntent.putExtra(EXTRA_IS_SETUP_FLOW, false);
        assertThat(new EnrollmentRequest(falseIntent, mContext, true).isSuw()).isFalse();
        assertThat(new EnrollmentRequest(falseIntent, mContext, false).isSuw()).isFalse();
    }

    @Test
    public void testIsAfterSuwOrSuwSuggestedAction() {
        // Default false
        assertThat(new EnrollmentRequest(new Intent(), mContext, true)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();
        assertThat(new EnrollmentRequest(new Intent(), mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent deferredTrueIntent = new Intent();
        deferredTrueIntent.putExtra(EXTRA_IS_DEFERRED_SETUP, true);
        assertThat(new EnrollmentRequest(deferredTrueIntent, mContext, true)
                .isAfterSuwOrSuwSuggestedAction()).isTrue();
        assertThat(new EnrollmentRequest(deferredTrueIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent deferredFalseIntent = new Intent();
        deferredFalseIntent.putExtra(EXTRA_IS_DEFERRED_SETUP, false);
        assertThat(new EnrollmentRequest(deferredFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();
        assertThat(new EnrollmentRequest(deferredFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent portalTrueIntent = new Intent();
        portalTrueIntent.putExtra(EXTRA_IS_PORTAL_SETUP, true);
        assertThat(new EnrollmentRequest(portalTrueIntent, mContext, true)
                .isAfterSuwOrSuwSuggestedAction()).isTrue();
        assertThat(new EnrollmentRequest(portalTrueIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent portalFalseIntent = new Intent();
        portalFalseIntent.putExtra(EXTRA_IS_PORTAL_SETUP, false);
        assertThat(new EnrollmentRequest(portalFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();
        assertThat(new EnrollmentRequest(portalFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent suggestedTrueIntent = new Intent();
        suggestedTrueIntent.putExtra(EXTRA_IS_SUW_SUGGESTED_ACTION_FLOW, true);
        assertThat(new EnrollmentRequest(suggestedTrueIntent, mContext, true)
                .isAfterSuwOrSuwSuggestedAction()).isTrue();
        assertThat(new EnrollmentRequest(suggestedTrueIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();

        final Intent suggestedFalseIntent = new Intent();
        suggestedFalseIntent.putExtra(EXTRA_IS_SUW_SUGGESTED_ACTION_FLOW, false);
        assertThat(new EnrollmentRequest(suggestedFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();
        assertThat(new EnrollmentRequest(suggestedFalseIntent, mContext, false)
                .isAfterSuwOrSuwSuggestedAction()).isFalse();
    }

    @Test
    public void testGetSuwExtras_inSuw() {
        final Intent suwIntent = new Intent();
        suwIntent.putExtra(EXTRA_IS_SETUP_FLOW, true);
        final EnrollmentRequest setupRequest = new EnrollmentRequest(suwIntent, mContext, true);

        final Bundle bundle = setupRequest.getSuwExtras();
        assertThat(bundle).isNotNull();
        assertThat(bundle.size()).isAtLeast(1);
        assertThat(bundle.getBoolean(EXTRA_IS_SETUP_FLOW)).isTrue();
    }

    @Test
    public void testGetSuwExtras_notInSuw() {
        final Intent suwIntent = new Intent();
        suwIntent.putExtra(EXTRA_IS_SETUP_FLOW, true);
        final EnrollmentRequest setupRequest = new EnrollmentRequest(suwIntent, mContext, false);

        final Bundle bundle = setupRequest.getSuwExtras();
        assertThat(bundle).isNotNull();
        assertThat(bundle.size()).isEqualTo(0);
    }

    @Test
    public void testIsSkipIntro() {
        // Default false
        assertThat(new EnrollmentRequest(new Intent(), mContext, true).isSkipIntro()).isFalse();
        assertThat(new EnrollmentRequest(new Intent(), mContext, false).isSkipIntro()).isFalse();

        final Intent trueIntent = new Intent();
        trueIntent.putExtra(EXTRA_SKIP_INTRO, true);
        assertThat(new EnrollmentRequest(trueIntent, mContext, true).isSkipIntro()).isTrue();
        assertThat(new EnrollmentRequest(trueIntent, mContext, false).isSkipIntro()).isTrue();

        final Intent falseIntent = new Intent();
        falseIntent.putExtra(EXTRA_SKIP_INTRO, false);
        assertThat(new EnrollmentRequest(falseIntent, mContext, false).isSkipIntro()).isFalse();
        assertThat(new EnrollmentRequest(falseIntent, mContext, false).isSkipIntro()).isFalse();
    }

    @Test
    public void testIsSkipFindSensor() {
        // Default false
        assertThat(new EnrollmentRequest(new Intent(), mContext, true).isSkipFindSensor())
                .isFalse();
        assertThat(new EnrollmentRequest(new Intent(), mContext, false).isSkipFindSensor())
                .isFalse();

        final Intent trueIntent = new Intent();
        trueIntent.putExtra(EXTRA_SKIP_FIND_SENSOR, true);
        assertThat(new EnrollmentRequest(trueIntent, mContext, true).isSkipFindSensor()).isTrue();
        assertThat(new EnrollmentRequest(trueIntent, mContext, false).isSkipFindSensor()).isTrue();

        final Intent falseIntent = new Intent();
        falseIntent.putExtra(EXTRA_SKIP_FIND_SENSOR, false);
        assertThat(new EnrollmentRequest(falseIntent, mContext, false).isSkipFindSensor())
                .isFalse();
        assertThat(new EnrollmentRequest(falseIntent, mContext, false).isSkipFindSensor())
                .isFalse();
    }

}

File: tests/unit/src/com/android/settings/biometrics2/ui/model/EnrollmentRequestTest.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.settings.biometrics2.ui.model

import android.content.Context
import android.content.Intent
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.android.settings.biometrics.BiometricEnrollActivity
import com.google.android.setupcompat.util.WizardManagerHelper
import com.google.common.truth.Truth
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class EnrollmentRequestTest {

    private val context = ApplicationProvider.getApplicationContext<Context>()

    @Test
    fun testIsSuw() {
        // Default false
        Truth.assertThat(EnrollmentRequest(Intent(), context, true).isSuw).isFalse()
        Truth.assertThat(EnrollmentRequest(Intent(), context, false).isSuw).isFalse()
        val trueIntent = Intent()
        trueIntent.putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, true)
        Truth.assertThat(EnrollmentRequest(trueIntent, context, true).isSuw).isTrue()
        Truth.assertThat(EnrollmentRequest(trueIntent, context, false).isSuw).isFalse()
        val falseIntent = Intent()
        trueIntent.putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, false)
        Truth.assertThat(EnrollmentRequest(falseIntent, context, true).isSuw).isFalse()
        Truth.assertThat(EnrollmentRequest(falseIntent, context, false).isSuw).isFalse()
    }

    @Test
    fun testIsAfterSuwOrSuwSuggestedAction() {
        // Default false
        Truth.assertThat(
            EnrollmentRequest(Intent(), context, true)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        Truth.assertThat(
            EnrollmentRequest(Intent(), context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val deferredTrueIntent = Intent()
        deferredTrueIntent.putExtra(WizardManagerHelper.EXTRA_IS_DEFERRED_SETUP, true)
        Truth.assertThat(
            EnrollmentRequest(deferredTrueIntent, context, true)
                .isAfterSuwOrSuwSuggestedAction
        ).isTrue()
        Truth.assertThat(
            EnrollmentRequest(deferredTrueIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val deferredFalseIntent = Intent()
        deferredFalseIntent.putExtra(WizardManagerHelper.EXTRA_IS_DEFERRED_SETUP, false)
        Truth.assertThat(
            EnrollmentRequest(deferredFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        Truth.assertThat(
            EnrollmentRequest(deferredFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val portalTrueIntent = Intent()
        portalTrueIntent.putExtra(WizardManagerHelper.EXTRA_IS_PORTAL_SETUP, true)
        Truth.assertThat(
            EnrollmentRequest(portalTrueIntent, context, true)
                .isAfterSuwOrSuwSuggestedAction
        ).isTrue()
        Truth.assertThat(
            EnrollmentRequest(portalTrueIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val portalFalseIntent = Intent()
        portalFalseIntent.putExtra(WizardManagerHelper.EXTRA_IS_PORTAL_SETUP, false)
        Truth.assertThat(
            EnrollmentRequest(portalFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        Truth.assertThat(
            EnrollmentRequest(portalFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val suggestedTrueIntent = Intent()
        suggestedTrueIntent.putExtra(WizardManagerHelper.EXTRA_IS_SUW_SUGGESTED_ACTION_FLOW, true)
        Truth.assertThat(
            EnrollmentRequest(suggestedTrueIntent, context, true)
                .isAfterSuwOrSuwSuggestedAction
        ).isTrue()
        Truth.assertThat(
            EnrollmentRequest(suggestedTrueIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        val suggestedFalseIntent = Intent()
        suggestedFalseIntent.putExtra(WizardManagerHelper.EXTRA_IS_SUW_SUGGESTED_ACTION_FLOW, false)
        Truth.assertThat(
            EnrollmentRequest(suggestedFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
        Truth.assertThat(
            EnrollmentRequest(suggestedFalseIntent, context, false)
                .isAfterSuwOrSuwSuggestedAction
        ).isFalse()
    }

    @Test
    fun testGetSuwExtras_inSuw() {
        val suwIntent = Intent()
        suwIntent.putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, true)
        val setupRequest = EnrollmentRequest(suwIntent, context, true)
        val bundle = setupRequest.suwExtras
        Truth.assertThat(bundle).isNotNull()
        Truth.assertThat(bundle.size()).isAtLeast(1)
        Truth.assertThat(bundle.getBoolean(WizardManagerHelper.EXTRA_IS_SETUP_FLOW)).isTrue()
    }

    @Test
    fun testGetSuwExtras_notInSuw() {
        val suwIntent = Intent()
        suwIntent.putExtra(WizardManagerHelper.EXTRA_IS_SETUP_FLOW, true)
        val setupRequest = EnrollmentRequest(suwIntent, context, false)
        val bundle = setupRequest.suwExtras
        Truth.assertThat(bundle).isNotNull()
        Truth.assertThat(bundle.size()).isEqualTo(0)
    }

    @Test
    fun testIsSkipIntro() {
        // Default false
        Truth.assertThat(EnrollmentRequest(Intent(), context, true).isSkipIntro).isFalse()
        Truth.assertThat(EnrollmentRequest(Intent(), context, false).isSkipIntro).isFalse()
        val trueIntent = Intent()
        trueIntent.putExtra(BiometricEnrollActivity.EXTRA_SKIP_INTRO, true)
        Truth.assertThat(EnrollmentRequest(trueIntent, context, true).isSkipIntro).isTrue()
        Truth.assertThat(EnrollmentRequest(trueIntent, context, false).isSkipIntro).isTrue()
        val falseIntent = Intent()
        falseIntent.putExtra(BiometricEnrollActivity.EXTRA_SKIP_INTRO, false)
        Truth.assertThat(EnrollmentRequest(falseIntent, context, false).isSkipIntro).isFalse()
        Truth.assertThat(EnrollmentRequest(falseIntent, context, false).isSkipIntro).isFalse()
    }

    @Test
    fun testIsSkipFindSensor() {
        // Default false
        Truth.assertThat(EnrollmentRequest(Intent(), context, true).isSkipFindSensor)
            .isFalse()
        Truth.assertThat(EnrollmentRequest(Intent(), context, false).isSkipFindSensor)
            .isFalse()
        val trueIntent = Intent()
        trueIntent.putExtra(EnrollmentRequest.EXTRA_SKIP_FIND_SENSOR, true)
        Truth.assertThat(EnrollmentRequest(trueIntent, context, true).isSkipFindSensor).isTrue()
        Truth.assertThat(EnrollmentRequest(trueIntent, context, false).isSkipFindSensor).isTrue()
        val falseIntent = Intent()
        falseIntent.putExtra(EnrollmentRequest.EXTRA_SKIP_FIND_SENSOR, false)
        Truth.assertThat(EnrollmentRequest(falseIntent, context, false).isSkipFindSensor)
            .isFalse()
        Truth.assertThat(EnrollmentRequest(falseIntent, context, false).isSkipFindSensor)
            .isFalse()
    }
}
